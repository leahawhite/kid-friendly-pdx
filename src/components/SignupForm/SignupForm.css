.signup-form {
  padding: 10px 0;
  text-align: left;
}

.signup-form input {
  border: 1px solid #aaa;
  display: block;
  margin-bottom: 10px;
  padding: 10px;
  text-align: left;
  width: 100%;
  border-radius: 4px;
}

.signup__hint {
  font-size: 16px;
  margin-bottom: 10px;
}

.password input {
  margin-bottom: 5px;
}

.signup-btn {
  border: 1px solid #aaa;
  background-color: #eee;
  border-radius: 4px;
  padding: 8px 12px;
  margin: 30px auto 0 auto;
  display: block;
}

input[defaultValue] {
  color: gray;
}