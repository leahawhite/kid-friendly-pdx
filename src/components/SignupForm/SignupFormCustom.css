.signup-form {
  padding: 40px 20px;
  text-align: left;
}

.signup-form input {
  border: 1px solid #ddd;
  display: block;
  font-size: 1.1em;
  margin-bottom: 10px;
  padding: 10px;
  text-align: left;
  width: 100%;
  border-radius: 4px;
}

.signup-btn {
  order: 1px solid #aaa;
  background-color: #eee;
  font-size: 1.1em;
  border-radius: 4px;
  padding: 8px 12px;
}

input:invalid {
  box-shadow: 0 0 5px 1px red;
}

input:focus:invalid {
  box-shadow: none;
}

input[defaultValue] {
  color: gray;
}