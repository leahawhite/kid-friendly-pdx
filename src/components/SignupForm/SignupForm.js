import React, { Component } from 'react';
import AuthApiService from '../../services/auth-api-service'
import './SignupForm.css';

export default class SignupForm extends Component {
  static defaultProps = {
    onSignUpSuccess: () => {}
  }

  constructor(props) {
    super(props);
    this.state = {
      error: null
    }
  }

  handleSubmit = event => {
    event.preventDefault();
    const { display_name, email, password } = event.target
    this.setState({ error: null })
    AuthApiService.signupUser({
      display_name: display_name.value, 
      email: email.value, 
      password: password.value 
    })
      .then(user => {
        display_name.value = ''
        email.value = ''
        password.value = ''
        this.props.onSignUpSuccess()
      })
      .catch(res => {
        this.setState({ error: res.error })
      })
  }
    
  render() {
    const { error } = this.state
    const hasError = error && <p className='red'>{error}</p>

    return (
      <form className="signup-form" onSubmit={this.handleSubmit}>
        <div role="alert">
          {hasError}
        </div>
        <div className="display_name">
          <label htmlFor="display_name">
            Display name 
            <span className="signup-rule"></span>
          </label>
          <input
            name="display_name"
            type="text"
            id="display_name"
            aria-label="display name"
            aria-required="true"
            placeholder="between 3 and 20 characters"
            required
            minLength="3"
            maxLength="20"
            autoComplete="off"
          />
        </div>
        <div className="email">
          <label htmlFor="email">
            Email address
          </label>
          <input
            name="email"
            type="email"
            id="email"
            aria-label="email address"
            aria-required="true" 
            required
            autoComplete="off"
          />
        </div>
        <div className="password">
          <label htmlFor="password">
            Password 
            <span className="signup-rule"></span>
          </label>
          <input
            name="password"
            type="password"
            id="password"
            placeholder="at least 6 characters, one number"
            aria-label="password"
            aria-required="true"
            minLength="6"
            maxLength="36"
            pattern=".*[0-9].*" 
            required
            autoComplete="off"
          />
        </div>
        <button type="submit" className="signup-btn" >
          Sign Up
        </button>
      </form>
    )
  }
}