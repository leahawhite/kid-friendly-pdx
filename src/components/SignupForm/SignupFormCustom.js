import React, { Component } from 'react';
import ValidationError from '../ValidationError';
import data from '../../data';
import './SignupFormCustom.css';

export default class SignupForm extends Component {
  static defaultProps = {
    onSignUpSuccess: () => {}
  }

  constructor(props) {
    super(props);
    this.state = {
      users: data.users,
      display_name: "",
      email: "",
      password: "",
      nameValid: false,
      emailValid: false,
      passwordValid: false,
      formValid: false,
      validationMessages: {
        display_name: "",
        email: "",
        password: "",
      },
    }
  }

  updateName = display_name => {
    this.setState({display_name}, () => {this.validateName(display_name)});
  }

  updateEmail = email => {
    this.setState({email}, () => {this.validateEmail(email)});
  }
  
  updatePassword = password => {
    this.setState({password}, () => {this.validatePassword(password)});
  }
  
  handleSubmit = event => {
    event.preventDefault();
    const { display_name, email, password } = this.state;
    console.log('Name: ', display_name);
    console.log('Email: ', email);
    console.log('Password: ', password);
    this.props.onSignUpSuccess(display_name, email, password);
  }

  validateName(fieldValue) {
    const { users } = this.state;
    const fieldErrors = {...this.state.validationMessages};
    let hasError = false;

    fieldValue = fieldValue.toString().trim();
    if(fieldValue.length === 0) {
      fieldErrors.display_name = 'Display name is required';
      hasError = true;
    } else {
      if (fieldValue.length < 3) {
        fieldErrors.display_name = 'Dispay name must be at least 3 characters long';
        hasError = true;
      } else {
        // if fieldValue exists in database already
        if (users.some(user => user.display_name.toLowerCase().includes(fieldValue.toLowerCase()))) {
          fieldErrors.display_name = 'Display name is already in use. Please choose another.';
          hasError = true;
        } else {
          fieldErrors.display_name = '';
          hasError = false;
        }
      }
    }
    this.setState({
      validationMessages: fieldErrors,
      nameValid: !hasError
    }, this.formValid );
  }

  validateEmail(fieldValue) {
    const fieldErrors = {...this.state.validationMessages};
    let hasError = false;

    fieldValue = fieldValue.toString().trim();
    if(fieldValue.length === 0) {
      fieldErrors.email = 'Email address is required';
      hasError = true;
    } else {
      if(!fieldValue.match(new RegExp(/^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/))) {
        fieldErrors.email = 'Email address is invalid';
        hasError = true;
      } else {
          fieldErrors.password = '';
          hasError = false;
        }
      }
    this.setState({
      validationMessages: fieldErrors,
      emailValid: !hasError
    }, this.formValid );
  }
  
  validatePassword(fieldValue) {
    const fieldErrors = {...this.state.validationMessages};
    let hasError = false;

    fieldValue = fieldValue.toString().trim();
    if(fieldValue.length === 0) {
      fieldErrors.password = 'Password is required';
      hasError = true;
    } else {
      if (fieldValue.length < 6 || fieldValue.length > 72) {
        fieldErrors.password = 'Password must be between 6 and 72 characters long';
        hasError = true;
      } else {
        if(!fieldValue.match(new RegExp(/^(?=.*[0-9])(?=.*[a-zA-Z])([a-zA-Z0-9]+)$/))) {
          fieldErrors.password = 'Password must contain at least one number and one letter';
          hasError = true;
        } else {
          fieldErrors.password = '';
          hasError = false;
        }
      }
    }
    this.setState({
      validationMessages: fieldErrors,
      passwordValid: !hasError
    }, this.formValid );
  }

  formValid() {
    this.setState({
      formValid: this.state.nameValid && this.state.emailValid && this.state.passwordValid
    })
  }
  
  render() {
    return(
      <>
        <form className="signup-form" onSubmit={e => this.handleSubmit(e)}>
            <div className="display_name">
              <label htmlFor="display_name">
                Display name
              </label>
              <input
                name="display_name"
                type="text"
                id="display_name"
                placeholder="Mr. Mister"
                required
                minLength="3"
                onChange={e => this.updateName(e)} />
              <ValidationError hasError={this.state.nameValid} message={this.state.validationMessages.display_name} />
            </div>
            <div className="email">
              <label htmlFor="email">
                Email address
              </label>
              <input
                name="email"
                type="email"
                id="email" 
                placeholder="mrmr@mister.com"
                required
                onChange={e => this.updateEmail(e)} />
              <ValidationError hasError={this.state.emailValid} message={this.state.validationMessages.email} />
            </div>
            <div className="password">
              <label htmlFor="password">
                Password
              </label>
              <input
                name="password"
                type="password"
                id="password"
                minLength="6"
                maxLength="72"
                pattern="(?=.*[0-9])" 
                required
                onChange={e => this.updatePassword(e)}/>
              <ValidationError hasError={this.state.passwordValid} message={this.state.validationMessages.password} />
              <div className="registration__hint">6 to 72 characters, must include a number</div>
            </div>
            <button type="submit" className="signup-btn" disabled={!this.state.formValid} >
              Sign Up
            </button>
          </form>
        </>
    )
  }
}