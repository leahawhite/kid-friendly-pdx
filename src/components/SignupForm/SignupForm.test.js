import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import SignupForm from './SignupForm';

describe('MapContainer component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<SignupForm />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders UI as expected', () => {
    const wrapper = shallow(<SignupForm />)
    expect(toJson(wrapper)).toMatchSnapshot()
  });
})