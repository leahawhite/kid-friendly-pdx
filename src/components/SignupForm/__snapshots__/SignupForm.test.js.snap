// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm component renders UI as expected 1`] = `
<form
  className="signup-form"
  onSubmit={[Function]}
>
  <div
    role="alert"
  />
  <div
    className="display_name"
  >
    <label
      htmlFor="display_name"
    >
      Display name
      <span
        className="signup-rule"
        id="signup-rule"
      >
         -- between 3 and 20 characters
      </span>
    </label>
    <input
      aria-describedby="signup-rule"
      aria-label="display name"
      aria-required="true"
      autoComplete="off"
      id="display_name"
      maxLength="20"
      minLength="3"
      name="display_name"
      required={true}
      type="text"
    />
  </div>
  <div
    className="email"
  >
    <label
      htmlFor="email"
    >
      Email address
    </label>
    <input
      aria-label="email address"
      aria-required="true"
      autoComplete="off"
      id="email"
      name="email"
      required={true}
      type="email"
    />
  </div>
  <div
    className="password"
  >
    <label
      htmlFor="password"
    >
      Password
      <span
        className="signup-rule"
        id="signup-rule"
      >
         -- at least 6 characters, one number
      </span>
    </label>
    <input
      aria-describedby="signup-rule"
      aria-label="password"
      aria-required="true"
      autoComplete="off"
      id="password"
      maxLength="36"
      minLength="6"
      name="password"
      pattern=".*[0-9].*"
      required={true}
      type="password"
    />
  </div>
  <button
    aria-label="sign up"
    className="signup-btn"
    type="submit"
  >
    Sign Up
  </button>
</form>
`;
